stdlib/asm/math/u64.masm

```diff
@@ -673,6 +676,32 @@ export.unchecked_shr
     cswap
 end
 
+# Performs left shift of one unsigned 64-bit integer preserving the overflow and
+# using the pow2 operation.
+# The input value to be shifted is assumed to be represented using 32 bit limbs.
+# The shift value is assumed to be in the range [0, 64).
+# Stack transition looks as follows:
+# [b, a_hi, a_lo, ...] -> [d_hi, d_lo, c_hi, c_lo, ...], where (d,c) = a << b, 
+# which d contains the bits shifted out.
+# This takes 20 cycles.
+export.overflowing_shl
+    pow2
+    u32split
+    exec.overflowing_mul
+end
+
+# Performs right shift of one unsigned 64-bit integer preserving the overflow and
+# using the pow2 operation.
+# The input value to be shifted is assumed to be represented using 32 bit limbs.
+# The shift value is assumed to be in the range [0, 64).
+# Stack transition looks as follows:
+# [b, a_hi, a_lo, ...] -> [d_hi, d_lo, c_hi, c_lo, ...], where c = a / 2^b, d = a % 2^b.
+export.overflowing_shr
```

Comment:
A more efficient way to do this could be to compute `c = a >> b` and `d = a << (64 - b)`. This would probably take around 50 cycles which should be better than `divmod` which I believe takes 60+ cycles.

On the other hand, maybe we'll find a way to optimize `divmod` in the future to make it faster. So, basically, I'm fine with either implementation, but we should probably leave a comment here mentioning the alternative.